gemma/gemma-inl.h

```diff
@@ -312,21 +312,29 @@ class GemmaAttention {
                 const size_t interleaved_idx = task / layer_config_.kv_heads;
                 const size_t query_idx = interleaved_idx % num_queries_;
                 const size_t batch_idx = interleaved_idx / num_queries_;
-                const size_t pos = queries_pos_[query_idx] + batch_idx;
+                size_t pos = queries_pos_[query_idx] + batch_idx;
                 const size_t cache_pos = div_seq_len_.Remainder(pos);
                 const size_t kv_offset = cache_pos * cache_pos_size_ +
                                          layer_ * cache_layer_size_ +
                                          head * layer_config_.qkv_dim * 2;
                 KVCache& kv_cache = kv_caches_[query_idx];
+
+                const size_t grp_size = layer_config_.grp_size;
+                const size_t ngb_size = layer_config_.ngb_size;
+                const bool self_extend = layer_config_.self_extend;
+
                 float* HWY_RESTRICT kv = kv_cache.kv_cache.get() + kv_offset;
                 const float* HWY_RESTRICT mha_kv =
                     activations_.q.Batch(interleaved_idx) + head * q_stride_ +
                     layer_config_.qkv_dim;
 
+                // When embedding position, we will use grouped key position
+                if (self_extend && pos > ngb_size) {
+                  pos /= grp_size;
```

Comment:
This might be expensive. Would you like to try precomputing and passing around a const hwy::Divisor& like we do with div_seq_len, see if that helps speed?